paddle/fluid/eager/backward.cc

```diff
@@ -74,14 +82,189 @@ std::unordered_map<GradNodeBase*, int> getInDegreeMap(
+        GradNodeBase* next_node = edge.GetMutableGradNode().get();
+
+        // Next node could be nullptr if it is leaf tensor with no
+        // AccumulationNode attached
+        // Or it could also originated from dispensable inputs
+        if (!next_node) continue;
+
+        // if node not in input_target_nodes,
+        // all the next_nodes of current node will be inserted to
+        // potential_stop_node
+        if (is_potential_stop_nodes) {
+          potential_stop_nodes->emplace(next_node);
+        }
+
+        // Update in_degree
+        if (!node_in_degree_map.count(next_node))
+          node_in_degree_map[next_node] = 0;
+        node_in_degree_map[next_node]++;
+
+        // Record depending relationship
+        (*depending_nodes)[next_node] = node;
+        queue.push(next_node);
+      }
+    }
+  }
+  // Update Graph Info, remove some stop_node in potential_stop_nodes
+  UpdateGraphInfo(target_nodes_inputmeta_map, depending_nodes,
+                  potential_stop_nodes);
+}
+
+void GetTargetNodesInfo(const std::vector<paddle::experimental::Tensor>& inputs,
+                        std::unordered_map<GradNodeBase*, AutogradMeta*>*
+                            target_nodes_inputmeta_map) {
+  VLOG(6) << "Get TargetNodes Info";
+  if (!inputs.empty()) {
+    VLOG(6) << "Inputs are not empty";
+    size_t num_inputs = inputs.size();
+    for (size_t i = 0; i < num_inputs; i++) {
+      AutogradMeta* auto_grad_meta =
+          EagerUtils::unsafe_autograd_meta(inputs[i]);
+      auto target_node = auto_grad_meta->GetMutableGradNode().get();
+
+      PADDLE_ENFORCE_NOT_NULL(target_node,
+                              paddle::platform::errors::Fatal(
+                                  "There is no grad op for input:%d or it's"
+                                  "stop_gradient=True",
+                                  i));
+      (*target_nodes_inputmeta_map)[target_node] = auto_grad_meta;
+    }
+  }
+}
+
+std::vector<paddle::experimental::Tensor> GetResults(
+    const std::vector<paddle::experimental::Tensor>& inputs,
+    const std::unordered_map<GradNodeBase*, paddle::experimental::Tensor>&
+        results_map,
+    bool allow_unused, bool create_graph) {
+  VLOG(6) << "Get Results";
+  if (inputs.empty()) return {};
+
+  std::vector<paddle::experimental::Tensor> results;
+  results.reserve(inputs.size());
+  auto results_map_ = results_map;
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    auto& input = inputs[i];
+    AutogradMeta* auto_grad_meta = EagerUtils::unsafe_autograd_meta(input);
+    auto target_node = auto_grad_meta->GetMutableGradNode().get();
+
+    if (results_map_.find(target_node) != results_map_.end()) {
+      // set StopGradient = !create_graph
+      auto result = results_map_[target_node];
+      AutogradMeta* tensor_auto_grad_meta = EagerUtils::autograd_meta(&result);
+      tensor_auto_grad_meta->SetStopGradient(!create_graph);
+      results.emplace_back(result);
```

Comment:
Just relax and perform a copy construction here, so you dont have to copy the entire map:

<img width="449" alt="image" src="https://user-images.githubusercontent.com/22334008/158095524-fcef4921-28f1-45e5-98a5-4222bfa1b421.png">

Reply:
Thx, use the input results_map instead of copying the entire map.